tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,685 @@
+    }
+    if (!(predictions instanceof Variable) && (!tf.scope().env().isEager())) {
+      // TODO output = backtrackIdentity(output); doesn't seem to work with Java version.
+      if (predictions.op().type().equals("Softmax")) {
+        if (predictions.op().numOutputs() != 1)
+          throw new IllegalArgumentException("output can only have 1 output");
+        predictions = predictions.op().output(0);
+        return tf.nn.softmaxCrossEntropyWithLogits(tLabels, predictions, -1);
+      }
+    }
+    Operand<T> one = tf.dtypes.cast(tf.constant(1), dataType);
+    Operand<T> epsilonConst = tf.dtypes.cast(tf.constant(EPSILON), dataType);
+    Operand<T> oneMinusEpsilonConst = tf.math.sub(one, epsilonConst);
+    predictions =
+        tf.math.div(
+            predictions, tf.reduceSum(predictions, tf.constant(axis), ReduceSum.keepDims(true)));
+    predictions = tf.clipByValue(predictions, epsilonConst, oneMinusEpsilonConst);
+
+    // Compute cross entropy from probabilities.
+    Operand<T> cce =
+        tf.reduceSum(
+            tf.math.mul(tLabels, tf.math.log(predictions)),
+            tf.constant(axis),
+            ReduceSum.keepDims(false));
+    return tf.math.neg(cce);
+  }
+
+  /**
+   * Computes the categorical hinge loss between labels and predictions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param labels true targets,  values are expected to be 0 or 1.
+   * @param predictions the predictions
+   * @param <T> the data type of the predictions and labels
+   * @return the categorical hinge loss
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> categoricalHinge(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    DataType<T> dataType = predictions.asOutput().dataType();
+    Operand<T> tLabels = tf.dtypes.cast(labels, dataType);
+    LossTuple<T> lossTuple = LossesImpl.squeezeOrExpandDimensions(tf, tLabels, predictions, null);
+    predictions = lossTuple.getTarget();
+    tLabels = lossTuple.getLabels();
+    Operand<T> one = tf.dtypes.cast(tf.constant(1), dataType);
+    Operand<T> zero = tf.dtypes.cast(tf.constant(0), dataType);
+
+    Operand<T> pos =
+        tf.reduceSum(
+            tf.math.mul(tLabels, predictions), tf.constant(-1), ReduceSum.keepDims(Boolean.FALSE));
+    Operand<T> neg =
+        tf.reduceMax(
+            tf.math.mul(tf.math.sub(one, tLabels), predictions),
+            tf.constant(-1),
+            ReduceMax.keepDims(Boolean.FALSE));
+    Operand<T> sub = tf.math.sub(neg, pos);
+    Operand<T> add = tf.math.add(sub, one);
+    return tf.math.maximum(zero, add);
+  }
+
+  /**
+   * Computes the cosine similarity loss between labels and predictions.
+   *
+   * <p>Note that it is a number between -1 and 1. When it is a negative number between -1 and 0, 0
+   * indicates orthogonality and values closer to -1 indicate greater similarity. The values closer
+   * to 1 indicate greater dissimilarity. This makes it usable as a loss function in a setting where
```

Comment:
The JavaDoc on the class should have matched the JavaDoc on `Losses.cosineSimilarity` method which returns values between -1 and 1. I have fixed the JavaDoc on the class `CosineSimilarity` to match the method JavaDoc.

Reply:
There is a discrepancy between the Python doc for `CosineSimilarity ` and the loss function `cosine_similarity`. Though the doc on `CosineSimilarity` indicates a value in the range of `[-1 to  0]`, the actual method indicates a range of `[-1 to  1]`.
There is nothing in the code between the two, that acts differently as `CosineSimilarity` just calls `cosine_similarity`.

The original `Keras` documentation on `CosineSimilarity`, agrees with the` [-1  to  1]` range. I am guessing the TensorFlow `CosineSimilarity` documentation is wrong.

Another weirdness creeps in in `metrics.CosineSimilarity`, which ends up calling another function called `cosine_proximity`. While the two different functions are in a similar category, they actually compute differently.
We'll address that issue in the `metrics` PR, as perhaps the metrics class should be renamed to `CosineProximity`.

Reply:
We should probably open an issue on the main TF repo flagging the docs issue, but that can wait.

Reply:
This javadoc is better, but I think it should mention that this function is inverted from the regular cosine similarity, as that's 1 when the values are most similar and -1 when they point in opposite directions. It makes sense that it is inverted because then you can minimise it sensibly, but it is confusing if you're just browsing through.